src/lazy/text/buffer.rs

```diff
@@ -0,0 +1,732 @@
+    ///
+    /// Note that `offset` is relative to the beginning of the buffer, not the beginning of the
+    /// larger stream of which the buffer is a piece.
+    pub fn slice_to_end(&self, offset: usize) -> TextBufferView<'data> {
+        TextBufferView {
+            data: &self.data[offset..],
+            offset: self.offset + offset,
+        }
+    }
+
+    /// Returns a slice containing all of the buffer's bytes.
+    pub fn bytes(&self) -> &[u8] {
+        self.data
+    }
+
+    /// Returns the number of bytes between the start of the original input byte array and the
+    /// subslice of that byte array that this `TextBufferView` represents.
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
+
+    /// Returns the number of bytes in the buffer.
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Returns `true` if there are no bytes in the buffer. Otherwise, returns `false`.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    pub fn match_whitespace(self) -> IonMatchResult<'data> {
+        is_a(WHITESPACE_CHARACTERS_AS_STR)(self)
+    }
+
+    /// Always succeeds and consumes none of the input. Returns an empty slice of the buffer.
+    // This method is useful for parsers that need to match an optional construct but don't want
+    // to return an Option<_>. For an example, see its use in `match_optional_whitespace`.
+    fn match_nothing(self) -> IonMatchResult<'data> {
+        // Use nom's `success` parser to return an empty slice from the head position
+        success(self.slice(0, 0))(self)
+    }
+
+    /// Matches zero or more whitespace characters.
+    pub fn match_optional_whitespace(self) -> IonMatchResult<'data> {
+        // Either match whitespace and return what follows or just return the input as-is.
+        // This will always return `Ok`, but it is packaged as an IonMatchResult for compatability
+        // with other parsers.
+        alt((Self::match_whitespace, Self::match_nothing))(self)
+    }
+
+    /// Matches a single top-level scalar value, the beginning of a container, or an IVM.
+    pub fn match_top_level(self) -> IonParseResult<'data, RawStreamItem<'data, TextEncoding>> {
+        let (remaining, value) = match self.match_value() {
+            Ok(value) => value,
+            Err(e) => return Err(e),
+        };
+
+        // TODO: Augment this method to take an `is_complete` flag that indicates whether the absence
+        //       of further values should return an `Incomplete` or a `RawStreamItem::EndOfStream`.
+
+        // TODO: Check to see if `value` is actually an IVM.
+        //       => If it's a symbol, try the IVM parser on it and see if it succeeds.
+        //       For now, we just return the value.
+        Ok((remaining, RawStreamItem::Value(value)))
+    }
+
+    /// Matches a single scalar value or the beginning of a container.
+    pub fn match_value(self) -> IonParseResult<'data, LazyRawTextValue<'data>> {
```

Comment:
(Question) Should this be called `read_value` as it returns a parsed value? Same for `match_top_level`?

Reply:
I was reserving the `read_*` names for methods that produce one of the `RawValueRef::*` variants. The `LazyRawTextValue` that this returns has its own `read()` method that serves that purpose.